### PR TITLE
X509 extension type admission #1601

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
##### Checklist
- [ ] documentation is added or updated
- [ ] tests are added or updated

##### Description of change
PR for issue #1601.

### CODE REVIEW

File: crypto/x509v3/v3_admis.c [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    ASN1_OPT(ADMISSION_SYNTAX, admissionAuthority, GENERAL_NAME),
+    ASN1_SEQUENCE_OF(ADMISSION_SYNTAX, contentsOfAdmissions, ADMISSIONS),
+} ASN1_SEQUENCE_END(ADMISSION_SYNTAX)
+
+IMPLEMENT_ASN1_FUNCTIONS(NAMING_AUTHORITY);
+IMPLEMENT_ASN1_FUNCTIONS(PROFESSION_INFO);
+IMPLEMENT_ASN1_FUNCTIONS(ADMISSIONS);
+IMPLEMENT_ASN1_FUNCTIONS(ADMISSION_SYNTAX);
+
+//DECLARE_STACK_OF(DIRECTORYSTRING)
```
Comment (FdaSilvaYY) [inline crypto/x509v3/v3_admis.c]: was a test ;)

File: crypto/x509v3/v3_admis.c [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+
+static int i2r_NAMING_AUTHORITY(const struct v3_ext_method *method, void *in,
+                                BIO *bp, int ind) 
+{
+    NAMING_AUTHORITY * namingAuthority = (NAMING_AUTHORITY*) in;
+
+    if (namingAuthority == NULL || namingAuthority->namingAuthorityId == NULL
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: break into two lines, so it looks more like common/parallel structure.
and indent the other parts of the || test an extra tabstop.

File: crypto/x509v3/v3_admis.c [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                                BIO *bp, int ind) 
+{
+    NAMING_AUTHORITY * namingAuthority = (NAMING_AUTHORITY*) in;
+
+    if (namingAuthority == NULL || namingAuthority->namingAuthorityId == NULL 
+        || namingAuthority->namingAuthorityText == NULL 
+        || namingAuthority->namingAuthorityUrl == NULL)
+        return 0;
+
+    if (BIO_printf(bp, "%*snamingAuthority: ", ind, "") <= 0) goto err;
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: newline missing.  and the other goto err's as well.

File: crypto/x509v3/v3_admis.c [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    if (namingAuthority->namingAuthorityId) {
+        char objbuf[128];
+        const char *ln = OBJ_nid2ln(OBJ_obj2nid(namingAuthority->namingAuthorityId));
+
+        if (BIO_printf(bp, "%*s  admissionAuthorityId: ", ind, "") <= 0) goto err;
+        OBJ_obj2txt(objbuf, sizeof objbuf, namingAuthority->namingAuthorityId, 1);
+
+        if (BIO_printf(bp, "%s%s%s%s\n", ln ? ln : "",
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: do other print routines use three ?: operators as well?

File: crypto/x509v3/v3_admis.c [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        if (BIO_printf(bp, "%*s  namingAuthorityText: ", ind, "") <= 0) goto err;
+        if (ASN1_STRING_print(bp, namingAuthority->namingAuthorityText) <= 0) goto err;
+        if (BIO_printf(bp, "\n") <= 0) goto err;
+    }
+    if (namingAuthority->namingAuthorityUrl != NULL ) {
+        if (BIO_printf(bp, "%*s  namingAuthorityUrl: ", ind, "") <= 0) goto err;
+        if (ASN1_STRING_print(bp, namingAuthority->namingAuthorityUrl) <= 0) goto err;
+        if (BIO_printf(bp, "\n") <= 0) goto err;
+    }
+
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: put the blank line after the return.  swap 90 and 91.

File: crypto/x509v3/v3_admis.c [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                    OBJ_obj2txt(objbuf, sizeof(objbuf), o, 1);
+                    if (BIO_printf(bp, "%*s      %s%s%s%s\n", ind, "", 
+                                   ln ? ln : "", ln ? " (" : "",
+                                   objbuf, ln ? ")" : "") <= 0) 
+                        goto err;
+                }
+            }
+        }
+    }
+
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: swap 169 and 170

File: crypto/x509v3/v3_admis.c [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    /* .ext_nid = */ NID_x509ExtAdmission,
+    /* .ext_flags = */ 0,
+    /*.it = */  ASN1_ITEM_ref(ADMISSION_SYNTAX),
+    NULL, NULL, NULL, NULL,
+    /*.i2s = */ NULL,
+    /*.s2i = */ NULL,
+    /*.i2v = */ NULL,
+    /*.v2i = */ NULL,
+    /*.i2r = */ &i2r_ADMISSION_SYNTAX,
+    /*.r2i = */ NULL,
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: what's that crazy comment syntax? :)  put the field name *after* the field value, or leave it off completely.
leave off trailing NULL elements.  and leave off trailing comma.
look at ocsp/v3_ocsp.c for formatting example.

File: crypto/x509v3/v3_admis.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+#ifndef HEADER_V3_ADMISSION_H
+# define HEADER_V3_ADMISSION_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+# include <openssl/x509.h>
+
+typedef struct NamingAuthority_st NAMING_AUTHORITY;
```
Comment (richsalz) [inline crypto/x509v3/v3_admis.h]: you can merge the typedef and the struct into one declaration.

File: crypto/x509v3/v3_admis.c (hunk 1 above)
Comment (richsalz) [inline crypto/x509v3/v3_admis.c]: what the C++ comment marker? :)

File: crypto/x509v3/v3_admis.c [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+static int i2r_NAMING_AUTHORITY(const struct v3_ext_method *method, void *in,
+                                BIO *bp, int ind) 
+{
+    NAMING_AUTHORITY * namingAuthority = (NAMING_AUTHORITY*) in;
+
+    if (namingAuthority == NULL || namingAuthority->namingAuthorityId == NULL 
+        || namingAuthority->namingAuthorityText == NULL 
+        || namingAuthority->namingAuthorityUrl == NULL)
+        return 0;
+
```
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: Same problem I mentioned before. You don't allow any field to be NULL (but the ASN.1 allows it as these are OPTIONAL). But (see below)...

File: crypto/x509v3/v3_admis.c [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    if (BIO_printf(bp, "%*snamingAuthority: ", ind, "") <= 0) goto err;
+
+    if (namingAuthority->namingAuthorityId) {
```
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: Here you are checking if a field is NULL again despite the test above returning early if it is NULL. I suspect you want to check that at least one field is not NULL in which case the above test is wrong.

**Comment by snhenson** [comment]
General question. What is the aim of adding this extension? Do you want to be able to just produce a human readable version of the extension or to be able to machine parse it? If the latter some of the typedefs need to be in public header files and you also need appropriate accessors.

If you want to be able to generate and add the extension too then additional code will be needed.

File: crypto/x509v3/v3_admis.c (hunk 10 above)
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: Also syntax issue. Should be an explicit != NULL.

Comment (FdaSilvaYY) [inline crypto/x509v3/v3_admis.c]: there is two fields naming `namingAuthorityId`, in two different structure.
A glitch from original code.

**Comment by FdaSilvaYY** [comment]
@snhenson : where is a linking issue, i don't understand the cause.
Sometime about STACK_OF(ASN1_STRING) methods, and DIRECTORYSTRING ...

File: crypto/x509v3/v3_admis.c (hunk 3 above)
Comment (FdaSilvaYY) [inline crypto/x509v3/v3_admis.c]: I will fix the goto err; a bit later ...

**Comment by wolftobias** [comment]
What´s the difference between "human readable version of the extension or to be able to machine parse it", 
parsing means ofter to produce a human reable format?! I need to get the profession info + oid out of the admission extension and I want also be able to set it to a certificate without it.

**Comment by snhenson** [comment]
The human readable version if what you get from X509_print_ext, x509 -text et al. This is designed to be read by humans and may not be completely unambiguous, so attempting to parse the textual output has some caveats. The current PR does this.

The machine parseable version is the structure representing the extension itself. That can be processed programmatically and should be totally unambiguous. The structures returned would need accessors to achieve this. The current PR AFAICS does not allow this.

If you want to set the extension in a certificate via a config file e.g. usable by the "ca" utility then it needs a companion r2i function. That is not in the current PR either.

**Comment by snhenson** [comment]
@FdaSilvaYY you need to have definitions for any STACK_OF types you use. I notice you're using a STACK_OF(DIRECTORYSTRING) which doesn't currently exist in OpenSSL. So you'd need to add the definition in an appropriate place (e.g. x509v3.h) also the stack type has to macth the function. So for a STACK_OF(DIRECTORTRING) you use sk_DIRECTORSTRING_num etc.

**Comment by mattcaswell** [comment]
Ping @richsalz and @snhenson. Is this ok now? @FdaSilvaYY this is marked as WIP. Is that still the case?

**Comment by ramahmoo** [comment]
This issue seems fixed in Branch, will it be merged into master soon?

File: crypto/x509v3/v3_admis.c [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                                BIO *bp, int ind) 
+{
+    NAMING_AUTHORITY * namingAuthority = (NAMING_AUTHORITY*) in;
+
+    if (namingAuthority == NULL 
+        || namingAuthority->namingAuthorityId == NULL 
+        || namingAuthority->namingAuthorityText == NULL 
+        || namingAuthority->namingAuthorityUrl == NULL)
+        return 0;
+
```
Comment (snhenson) [inline crypto/x509v3/v3_admis.c]: This is still a problem. We're requiring _all_ fields are not NULL but the logic below and the ASN.1 module allows the fields to be omitted.

Comment (FdaSilvaYY) [inline crypto/x509v3/v3_admis.c]: oups, logic inversion.
I fixing it now.

**Comment by FdaSilvaYY** [comment]
I just rework the original patch to make it pushable.
For any additional feature like: 
- config file parsing and support 
- new methods/accessors to manipulate these objects.
Please fill an issue ;)

**Comment by wolftobias** [comment]
ping @FdaSilvaYY Is that now ready to get committed into master?

**Comment by matthiasballreich** [comment]
yes, is it ready to commit into master?
and @FdaSilvaYY are you able to wirte / include some more X509 Certificate Extensions which are not yet supported by OpenSSL by default, like SubjectDirectoryAttributes, QcStatements and some more Attributes like your AddmissionSyntax Attribute. I can provide some more infos if needed :)

**Comment by richsalz** [comment]
@snhenson does this look okay now?

**Comment by FdaSilvaYY** [comment]
Quick rebasse to last master.
@matthiasballreich : I barely have time to rebase my PR, may be later.

**Comment by richsalz** [comment]
looks good to me; ping @snhenson

**Comment by richsalz** [comment]
done.  thanks for your patience :)

**Comment by matthiasballreich** [comment]
I compiled the actual dev build, but why i can't use the ADMISSION_SYNTAX Data-Type?
NID_x509extadmission will be recognized in my Code but not ADMISSION_SYNTAX. Where is the problem?

**Comment by FdaSilvaYY** [comment]
Because the type structures declarations were not made into public header file ( i.e: inside `include/openssl` folder)

**Comment by matthiasballreich** [comment]
ahh yeah, You're right. I did not even think of that - my mistake. Thanks ;)

**Comment by wolftobias** [comment]
I have the same issue. Today I build the current master and the admission header is not available and therefore the admission structures can not be used by now.

**Comment by matthiasballreich** [comment]
How did i use the code correctly?

I'm using:
```
int loc = X509_get_ext_by_NID(CERT, ADMISSION_NID, -1);
X509_EXTENSION *ex = X509_get_ext(CERT, loc);
if(ex) {
    ADMISSION_SYNTAX* as = (ADMISSION_SYNTAX*) X509V3_EXT_d2i(ex);
    ....
}
```

But there is a problem, when i try to get the `admissionAuthority` which is a `GeneralName` form the main `ADMISSION_SYNTAX`
```
 AdmissionSyntax ::= SEQUENCE
     {
       admissionAuthority GeneralName OPTIONAL,
       contentsOfAdmissions SEQUENCE OF Admissions
     }
```
As value for admissionAuthority->type i got a value like 1366944 or something which is not correctly i mean. So have you tested the extension in a proper way or i'm using the wrong code?

**Pull request closed without merging**